Pr paddle integration

sfraczek opened the pull request:
This PR contains fixes and features required for integration with PaddlePaddle framework, small fix to a test and a benchmark for testing aeon's performance with config used for PaddlePaddle's image classification task for resnet50 on imagenet.

Its best to review it commit by commit.

AdrShv committed:
Deterministic shuffle images in block when reading from cache

Workaround to enable shuffle images in block when cache is turned off

AdrShv committed:
initial version

AdrShv committed:
fix bug

AdrShv committed:
fix bugs

AdrShv committed:
add unit tests

AdrShv committed:
clean up

AdrShv committed:
fix multinode shuffling bug

Commit:
added benchmark of imagenet for PaddlePaddle config

Commit:
add resize_short

Commit:
add resize_short

Commit:
add bgr_to_rgb

Commit:
fix test area_scale

Commit:
add normalization

Commit:
recover missing test/test_data/flowers.jpg

sfraczek commented:
@AdrShv @arogowie-intel Should I pass `bgr_to_rgb` from augmentation config to image config like is done for `fixed_aspect_ratio` in src/provider.cpp:156 rather than pass it to `image config`? 
**[EDIT] a change is ready I can push it if you agree**

Should I move normalization to `transformer` rather than doing it in `loader`?

arogowie-intel commented:
I think that `bgr_to_rgb` should be in image configuration since it is only change of image representation. Normalization in contrast to this changes image data, thus should be in bag of augmentations.

sfraczek commented:
On the other hand, why is fixed_aspect_ratio given through augmentation I wonder? What is the pattern here?

sfraczek commented:
@arogowie-intel I have moved normalization to augmentation config

Commit:
update docs

Commit:
rename normalize->standardize

Commit:
add missing test/test_data/img_2112_70.jpg

Commit:
fix warnings on member initializer list order

Commit:
fix bug

Commit:
add missing test.cpio

Commit:
update loader.deterministic test

Commit:
disable nds in tests

Commit:
review fix: make mean & stddev optional

Commit:
add benchmark for pure image reading in test/test_benchmark.cpp

AdrShv commented on src/etl_image.cpp:
```diff
@@ -150,7 +156,11 @@ cv::Mat image::transformer::transform_single_image(shared_ptr<augment::image::pa
+    // TODO(sfraczek): add test for this resize short
+    cv::Mat resizedShortImage;
+    image::resize_short(expandedImage, resizedShortImage, img_xform->resize_short_size);
+
```
Does it mean that we always resize short size ? as I remember we don't need to do that in training mode

sfraczek replied:
If resize_short_size is 0 then it doesn't resize. The check is inside this function. The default value is 0.

AdrShv replied:
ok, but please pull out this condition from resize_shot function

sfraczek replied:
done

AdrShv commented on src/etl_image.cpp:
```diff
@@ -203,51 +234,82 @@ void image::loader::load(const vector<void*>& outlist, shared_ptr<image::decoded
+    else
+    {
+        from_to = {0, 0};
+    }
```
what about other channels size, for instance 4 ?

sfraczek replied:
Its not handled. How should we handle that?  Copy 0->0, 1->1, 2->2, 3->3?

AdrShv replied:
I would to use loop approach from line 263

sfraczek replied:
done

AdrShv commented on src/etl_image.cpp:
```diff
@@ -203,51 +234,82 @@ void image::loader::load(const vector<void*>& outlist, shared_ptr<image::decoded
+                }
+                // channelwise call
+                image::standardize(channels, m_mean, m_stddev);
             }
```
what if we don't need normalize image ? As I can see, you always call that

sfraczek replied:
Inside this function there is one check and there is assumption that the check in line 223 https://github.com/NervanaSystems/aeon/pull/6/files#diff-a4ccfbf4e57f797ce3bbb8864ceaf6cbR223 was also performed. If mean/stddev are not provided then it returns without changing the image.

AdrShv replied:
ok, but please, pull out that check from standarize function

sfraczek replied:
@AdrShv should I move it outside?

AdrShv commented on src/image.cpp:
```diff
@@ -84,6 +84,55 @@ void image::resize(const cv::Mat& input, cv::Mat& output, const cv::Size2i& size
+            subtract(input[i], s_mean, input[i]);
+            // divide
+            multiply(input[i], s_stddev, input[i]);
+        }
```
4th element of s_mean is 0. How could OpenCV  know that you have 3th element vector ?? Does it work ?

sfraczek replied:
I check that if mean and stddev are used then the length of each must be equal to number of channels in https://github.com/NervanaSystems/aeon/pull/6/files#diff-a4ccfbf4e57f797ce3bbb8864ceaf6cbR223

AdrShv commented on src/image.cpp:
```diff
@@ -84,6 +84,55 @@ void image::resize(const cv::Mat& input, cv::Mat& output, const cv::Size2i& size
+    }
+
+    for (int i = 0; i < input.size(); i++)
+    {
```
please use for(auto& input_image: input) version of loop, that allow do code more readable, because you avoid write [i] everywhere

sfraczek replied:
I iterate over three arrays with this index in this loop so I think I can't.

AdrShv commented on src/image.cpp:
```diff
@@ -109,15 +158,7 @@ void image::convert_mix_channels(vector<cv::Mat>& source,
-        size_t size = target[0].total() * target[0].elemSize();
-        memcpy(target[0].data, (*prepared_source)[0].data, size);
-    }
-    else
```
why did you remove that optimization?

sfraczek replied:
I don't remember. I think I could bring it back if I add a check that from_to is  something in the pattern of 0,0 ;    0,0, 1,1 ;     0,0, 1,1, 2,2;     0,0, 1,1, 2,2, 3,3. 
I don't know if it will work in this case https://github.com/NervanaSystems/aeon/pull/6/files#diff-a4ccfbf4e57f797ce3bbb8864ceaf6cbR310 where the target image is a cropped with cv::Rect matrix. I think the data is not contiguous there?

sfraczek replied:
I could actually check  strides (step[0] and step[1]) and compare them with width and height of images

sfraczek replied:
I check by isContinuous

AdrShv commented on src/image.cpp:
```diff
@@ -84,6 +84,55 @@ void image::resize(const cv::Mat& input, cv::Mat& output, const cv::Size2i& size
+    {
+        // divide
+        multiply(input[i], 1. / 255., input[i]);
+
```
Please add cv::

sfraczek replied:
done

sfraczek reviewed (commented):
@AdrShv I've answered your questions

Commit:
resize short condition moved outside

Commit:
from_to uses loop and is moved to constructor and added const

Commit:
moved check out of standardize

Commit:
Refactored standardize to make it more readable

Commit:
rephrase cmoment

Commit:
bring back optimization to mix channels if is continuous

Commit:
added cv::

AdrShv commented on src/image.cpp:
```diff
@@ -84,9 +84,65 @@ void image::resize(const cv::Mat& input, cv::Mat& output, const cv::Size2i& size
+
+    auto resized_width  = static_cast<int>(std::round(input.cols * percent));
+    auto resized_height = static_cast<int>(std::round(input.rows * percent));
+    cv::resize(input, output, cv::Size2i(resized_width, resized_height), 0, 0, CV_INTER_LINEAR);
```
this code is almost the same as src/augment_image.cpp : line 193. May be we can create separate function to avoid copy past.

arogowie-intel commented on src/augment_image.cpp:
```diff
@@ -191,6 +193,16 @@ shared_ptr<augment::image::params> augment::image::param_factory::make_params(
+        {
+            auto percent =
+                static_cast<float>(resize_short_size) / std::min(input_width, input_height);
+            auto resized_width  = static_cast<int>(std::round(input_width * percent));
```
Why not to use `nervana::unbiased_round` ?

sfraczek replied:
its like in python

arogowie-intel commented on src/augment_image.cpp:
```diff
@@ -191,6 +193,16 @@ shared_ptr<augment::image::params> augment::image::param_factory::make_params(
         cv::Size2f out_shape(output_width * _horizontal_distortion, output_height);
 
+        // TODO(sfraczek): add test for this resize short
+        if (resize_short_size > 0)
```
@AdrShv Why are we doing this only when cropping is enabled?

sfraczek replied:
I don't know if we should do it for other paths as well. Maybe we should

arogowie-intel commented on src/etl_image.hpp:
```diff
@@ -170,15 +172,22 @@ class nervana::image::transformer
+    uint32_t            m_channels;
+    std::vector<double> m_mean;
+    std::vector<double> m_stddev;
+    std::vector<int>    from_to;
```
Please add `m_` prefix.

arogowie-intel commented on src/image.cpp:
```diff
@@ -109,7 +165,13 @@ void image::convert_mix_channels(vector<cv::Mat>& source,
 
-    if (prepared_source->size() == 1 && target.size() == 1)
+    bool index  = 0;
+    bool no_mix = std::find_if(from_to.begin(), from_to.end(), [&index](int element) {
```
This is extremely unreadable - what condition did you want to achieve?

Commit:
disk benchmarks

Commit:
Disable test instead of commenting them out.

Commit:
Add interpolation_method for augmentation config.

- Add interpolation parameter to image::resize function.

Commit:
Simplify condition when to mix channels.

Use passed value of image config bgr_to_rgb key.

Commit:
Add utility for reading binary files.

Commit:
Enable generating manifest file with provided set of images.

Commit:
Add UT checking for image provider augmentation and loading.

Commit:
Comment out UTs which cause compilation errors.

Commit:
Fix some errors, refactoring.

- Fix condition for not mixing channels.
- Remove `std::` prefix
- Remove where unnecessary `nervana::` prefix.
- Add `m_` prefix to `from_to` member.

Commit:
Disable test_loader.deterministic UT.

sfraczek commented on src/image.cpp:
```diff
@@ -105,18 +105,27 @@ void image::resize(const cv::Mat& input,
+    cv::resize(input,
+               output,
+               get_resized_short_size(input.cols, input.rows, target_size),
+               0, 0, CV_INTER_LINEAR);
```
Maybe we should consider using the interpolation provided by config

Commit:
Add missing input cpio test data.

sfraczek commented on test/test_loader.cpp:
```diff
@@ -460,7 +460,7 @@ static std::string generate_manifest_file(
 }
 
-TEST(loader, deterministic)
+TEST(DISABLED_loader, deterministic)
```
alternatively to disabling you could hardcode new values that you get after your changes but the question is what has caused the change in values.

Commit:
Add augmentation pipeline UT for validation config.

Commit:
Add reference results binary file for validation augmentation UT.

Commit:
Use interpolation method from config.

Commit:
Merge remote-tracking branch 'origin/ashvay/disk_benchmark' into pr-paddle-integration

AdrShv commented on src/image.cpp:
```diff
@@ -63,30 +90,90 @@ void image::add_padding(cv::Mat& input, int padding, cv::Size2i crop_offset)
+                   cv::Mat& output,
+                   const cv::Size2i& size,
+                   const string& inter_method)
 {
```
Why we can't pass to resize cv::InterpolationFlags instead of string ?

AdrShv replied:
ok, now I see why :)

Commit:
Fix conversion to cropbox cv::Rect.
